lib/utils.js

```diff
@@ -613,6 +613,37 @@ api.verifyCapabilityChain = async ({
       }
     }
 
+    // ensure that the delegation chain does not expand the `allowedAction` of
+    // the original delgation
+    const fullCapabilityChain = pushedGivenCapability ?
+      dereferencedCapabilities : [...dereferencedCapabilities, capability];
+    let parentAllowedAction;
+    for(const c of fullCapabilityChain) {
```

Comment:
Any reason not to put this check in CapabilityDelegation (around here: https://github.com/digitalbazaar/ocapld.js/blob/master/lib/CapabilityDelegation.js#L134)? It seems we could just check against the `verifiedParentCapability` in that case and it would keep all of these sorts of checks in one place.

Reply:
I didn't want to put it there because of issues related to whether or not the full chain was being checked.  I believe the loop you mentioned can not be relied upon the contain the full chain.

Reply:
> I believe the loop you mentioned can not be relied upon the contain the full chain.

That sounds like a problem that needs to be fixed rather than putting some checks here and other checks in CapabilityDelegation. I think all the checks should be in CapabilityDelegation (other than the ones that are obviously related to the root zcap only) and we should make sure the verification process doesn't have any gaps. If it does have gaps, then those checks on caveats and whatever else that happen in CapabilityDelegation would also be getting skipped, creating security holes.

Reply:
If all the checks can't be in one place for some reason, we need a better way to reason about what goes where. If they can be in one place, but that place needs to be in the verify capability chain function instead, we should move them all there.

Note that there's also the notion that we want to start at the root and verify everything as we go -- so we don't waste cycles verifying stuff at the tail of the chain if it fails before we get there. The only exception to this is the "inspectCapabilityChain" which runs at the end, because it is the heaviest process and likely needs to call out to a database, etc., so it makes sense to do it last (and it can be done more efficiently if done all at once).